meerk40t/gui/scenewidgets/rectselectwidget.py

```diff
@@ -162,51 +172,230 @@ def update_statusmsg(self, value):
+                    e.translated(dx, dy)
+            self.scene.context.elements.update_bounds([b[0] + dx, b[1] + dy, b[2] + dx, b[3] + dy])
+            self.scene.request_refresh()
+
         if modifiers is not None:
             self.modifiers = modifiers
 
         elements = self.scene.context.elements
         if event_type == "leftdown":
+            self.mouse_down_time = perf_counter()
+            self.mode = "unclear"
             self.start_location = space_pos
             self.end_location = space_pos
+            if contains(self.scene.context.elements._emphasized_bounds, space_pos):
+                self.can_drag_move = True
             # print ("RectSelect consumed leftdown")
             return RESPONSE_CONSUME
         elif event_type == "leftclick":
             # That's too fast
             # still chaining though
             self.scene.request_refresh()
-            self.start_location = None
-            self.end_location = None
+            self.reset()
             return RESPONSE_CHAIN
         elif event_type == "leftup":
-            if self.start_location is None:
-                return RESPONSE_CHAIN
-            _ = self.scene.context._
-            self.update_statusmsg(_("Status"))
-            elements.validate_selected_area()
-            sx = min(self.start_location[0], self.end_location[0])
-            sy = min(self.start_location[1], self.end_location[1])
-            ex = max(self.start_location[0], self.end_location[0])
-            ey = max(self.start_location[1], self.end_location[1])
-            self.rect_select(elements, sx, sy, ex, ey)
-
-            self.scene.request_refresh()
-            self.scene.context.signal("select_emphasized_tree", 0)
-
-            self.start_location = None
-            self.end_location = None
+            if self.mode == "select":
+                if self.start_location is None:
+                    return RESPONSE_CHAIN
+                _ = self.scene.context._
+                self.update_statusmsg(_("Status"))
+                elements.validate_selected_area()
+                sx = min(self.start_location[0], self.end_location[0])
+                sy = min(self.start_location[1], self.end_location[1])
+                ex = max(self.start_location[0], self.end_location[0])
+                ey = max(self.start_location[1], self.end_location[1])
+                self.rect_select(elements, sx, sy, ex, ey)
+
+                self.scene.request_refresh()
+                self.scene.context.signal("select_emphasized_tree", 0)
+            else:
+                def shortest_distance(p1, p2):
+                    """
+                    Calculates the shortest distance between two arrays of 2-dimensional points.
+                    """
+                    # Calculate the Euclidean distance between each point in p1 and p2
+                    dist = np.sqrt(np.sum((p1[:, np.newaxis] - p2) ** 2, axis=2))
```

Comment:
Somebody's getting fancy.

Reply:
Most points are actually complex numbers. You could just subtract them and take the absolute power.

So if a and b are points in complex numbers the Euclidean distance is `np.abs(a - b)` even groups of points should basically be like that too. This code man actually only need to be `np.abs(p1 - p2)` assuming we're dealing with complex numbers here which most points are.

It's notably sort of complex to get a grasp on, but worth the confusing since it makes things a lot easier.